src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.StringSegment.cs

```diff
@@ -363,6 +360,126 @@ private void WriteStringSegmentData(ReadOnlySpan<byte> escapedValue)
+
+                if (!_options.SkipValidation)
+                {
+                    ValidateWritingValue();
+                }
+
+                WriteStringSegmentPrologue();
+
+                PreviousSegmentEncoding = SegmentEncoding.Base64;
+                _tokenType = Utf8JsonWriter.StringSegmentSentinel;
+            }
+            else
+            {
+                ValidateEncodingDidNotChange(SegmentEncoding.Base64);
+            }
+
+            // The steps to write a string segment are to complete the previous partial code point
+            // and escape either of which might not be required so there is a fast path for each of these steps.
+            if (HasPartialCodePoint)
+            {
+                WriteBase64StringSegmentWithLeftover(value, isFinalSegment);
+            }
+            else
+            {
+                WriteBase64StringSegmentData(value, isFinalSegment);
+            }
+
+            if (isFinalSegment)
+            {
+                WriteStringSegmentEpilogue();
+
+                SetFlagToAddListSeparatorBeforeNextItem();
+                PreviousSegmentEncoding = SegmentEncoding.None;
+                _tokenType = JsonTokenType.String;
+            }
+        }
+
+        private void WriteBase64StringSegmentWithLeftover(scoped ReadOnlySpan<byte> bytes, bool isFinalSegment)
+        {
+            Debug.Assert(HasPartialCodePoint);
+            Debug.Assert(PreviousSegmentEncoding == SegmentEncoding.Base64);
+
+            scoped ReadOnlySpan<byte> partialCodePointBuffer = PartialBase64CodePoint;
+
+            Span<byte> combinedBuffer = stackalloc byte[3];
+            combinedBuffer = combinedBuffer.Slice(0, ConcatInto(partialCodePointBuffer, bytes, combinedBuffer));
+            if (combinedBuffer.Length is 3)
+            {
+                // Divide up the code point bytes into its own buffer and the remainder of the input buffer.
+                bytes = bytes.Slice(3 - partialCodePointBuffer.Length);
+                partialCodePointBuffer = combinedBuffer.Slice(0, 3);
+            }
+            else
+            {
+                Debug.Assert(combinedBuffer.Length is 1 or 2);
+                // Need more data. If this is a final segment, then the encoder will append '=' as needed.
+                Debug.Assert(bytes.Length + partialCodePointBuffer.Length < 3);
+                bytes = combinedBuffer;
+                partialCodePointBuffer = [];
+            }
+
+            // It doesn't matter if we pass true or false for isFinalSegment since we are guaranteed to not have a partial code point
+            // here (it is either empty or completed using the combined buffer above).
+            WriteBase64StringSegmentData(partialCodePointBuffer, false);
+
+            ClearPartialCodePoint();
+
+            WriteBase64StringSegmentData(bytes, isFinalSegment);
+        }
+
+        private void WriteBase64StringSegmentData(ReadOnlySpan<byte> bytes, bool isFinal)
```

Comment:
Nit: naming consistency:
```suggestion
        private void WriteBase64StringSegmentData(ReadOnlySpan<byte> bytes, bool isFinalSegment)
```